File: models/image.rb
require 'fileutils'
require 'open-uri'
require 'RMagick'

class Image

    @path
    @image

    # Image's constructor
    #
    # @param filename
    def initialize filename
        @path = Image._path filename
        self._load
    end

    # Resize the current image
    #
    # @param [Integer] width
    # @param [Integer] height
    def resize width, height
        @path = self._path_resized width, height
        if not File.exist? @path then
            width = self.width_resized width, height
            height = self.height_resized width, height
            @image = @image.resize width, height
            self._save
        else
            self._load
        end
    end

    # Crop the current image
    #
    # @param [Integer] x
    # @param [Integer] y
    # @param [Integer] width
    # @param [Integer] height
    def crop x, y, width, height
        @path = self._path_cropped x, y, width, height
        if not File.exist? @path then
            @image = @image.crop x, y, width, height
            self._save
        else
            self._load
        end
    end

    # Get the image width
    #
    # @return [Integer]
    def width
        return @image.columns
    end

    # Get the image height
    #
    # @return [Integer]
    def height
        return @image.rows
    end

    # Get the image width after resizing
    #
    # @param [Integer] width
    # @param [Integer] height
    # @return [Integer]
    def width_resized width, height
        ratio = @image.columns/@image.rows
        if ratio > width/height then
            width = @image.columns * height / @image.rows
        end
        return width
    end

    # Get the image height after resizing
    #
    # @param [Integer] width
    # @param [Integer] height
    # @return [Integer]
    def height_resized width, height
        ratio = @image.columns/@image.rows
        if ratio < width/height then
            height = @image.rows * width / @image.columns
        end
        return height
    end

    # Get the image path
    #
    # @return [String]
    def path
        return @path
    end

    # Get the image filename
    #
    # @return [String]
    def filename
        return File.basename @path
    end

    # Get the image type
    #
    # @return [String]
    def mime_type
        extension = File.extname @path
        type = extension.sub /\./, ''
        if type.eql? 'jpg' then
            type = 'jpeg'
        end
        return "image/#{type}"
    end

    # Upload a file
    #
    # @param [String] file
    # @return [Image]
    def self.upload file
        filename = Image._filename file[:filename]
        path = Image._path filename
        begin
            FileUtils.cp file[:tempfile].path, path
        rescue
            raise "Unable to upload this file"
        end
        return Image.new filename
    end

    # Download an URL
    #
    # @param [String] url
    # @return [Image]
    def self.download url
        filename = Image._filename url
        path = Image._path filename
        begin
            input = open url
            output = open path, 'wb'
            output.write input.read
            output.close
        rescue
            raise "Unable to fetch this url"
        end
        return Image.new filename
    end

    # Load the image
    def _load
        begin
            image = Magick::Image.read @path
            @image = image.first
        rescue
            raise "Unable to load this image"
        end
    end

    # Save the image on the filesystem
    def _save
        @image.write(@path){self.density="72x72"}
    end

    # Get the path for the resized image
    #
    # @param [Integer] width
    # @param [Integer] height
    # @return [String]
    def _path_resized width, height
        extension = File.extname @path
        basename = File.basename @path, extension
        return Image._path "#{basename}-#{width}x#{height}#{extension}"
    end

    # Get the path for the cropped image
    #
    # @param [Integer] x
    # @param [Integer] y
    # @return [String]
    def _path_cropped x, y, width, height
        extension = File.extname @path
        basename = File.basename @path, extension
        return Image._path "#{basename}-#{width}x#{height}-#{x}x#{y}#{extension}"
    end

    # Get the complete path for a given filename
    #
    # @param [String] filename
    # @return [String]
    def self._path filename
        return "#{File.dirname(__FILE__)}/../tmp/#{filename}"
    end

    # Generate a filename
    #
    # @param [String] path
    # @return [String]
    def self._filename path
        extension = File.extname path
        return "#{Time.now.to_i}#{extension}"
    end

end

File: app/Image.php
<?php

namespace App;

class Image
{
    private $_path;
    private $_image;

    /**
     * Image's constructor
     *
     * @param $filename
     */
    public function __construct($filename)
    {
        $this->_path = self::_path($filename);
        $this->_load();
    }

    /**
     * Resize the current image
     *
     * @param Integer $width
     * @param Integer $height
     */
    public function resize($width, $height)
    {
        $this->_path = $this->_pathResized($width, $height);
        if (!file_exists($this->_path)) {
            $width = $this->widthResized($width, $height);
            $height = $this->heightResized($width, $height);
            $this->_image->resize($width, $height);
            $this->_save();
        } else {
            $this->_load();
        }
    }

    /**
     * Crop the current image
     *
     * @param Integer $x
     * @param Integer $y
     * @param Integer $width
     * @param Integer] $height
     */
    public function crop($x, $y, $width, $height)
    {
        $this->_path = $this->_pathCropped($x, $y, $width, $height);
        if (!file_exists($this->_path)) {
            $this->_image->crop($width, $height, $x, $y);
            $this->_save();
        } else {
            $this->_load();
        }
    }

    /**
     * Get the image width
     *
     * @return Integer
     */
    public function width()
    {
        return $this->_image->width();
    }

    /**
     * Get the image height
     *
     * @return Integer
     */
    public function height()
    {
        return $this->_image->height();
    }

    /**
     * Get the image width after resizing
     *
     * @param Integer $width
     * @param Integer $height
     * @return Integer
     */
    public function widthResized($width, $height)
    {
        $ratio = $this->width() / $this->height();
        if ($ratio > $width/$height) {
            $width = $this->width() * $height / $this->height();
        }
        return $width;
    }

    /**
     * Get the image height after resizing
     *
     * @param Integer $width
     * @param Integer $height
     * @return Integer
     */
    public function heightResized($width, $height)
    {
        $ratio = $this->width() / $this->height();
        if ($ratio < $width/$height) {
            $height = $this->height() * $width / $this->width();
        }
        return $height;
    }

    /**
     * Get the image path
     *
     * @return String
     */
    public function path()
    {
        return $this->_path;
    }

    /**
     * Get the image filename
     *
     * @return String
     */
    public function filename()
    {
        return basename($this->_path);
    }

    /**
     * Upload a file
     *
     * @param String $file
     * @return Image
     */
    public static function upload($file)
    {
        if (!$file->isValid()) {
            throw new \Exception("Unable to upload this file");
        }
        $filename = self::_filename($file->getClientOriginalName());
        $path = self::_path($filename);
        $file->move(dirname($path), basename($path));
        return new self($filename);
    }

    /**
     * Download an URL
     *
     * @param String $url
     * @return Image
     */
    public static function download($url)
    {
        $filename = self::_filename($url);
        $path = self::_path($filename);
        try {
            $client = new \GuzzleHttp\Client();
            $data = $client->get($url)->getBody();
            file_put_contents($path, $data);
        } catch (\Exception $e) {
            throw new \Exception("Unable to fetch this url");
        }
        return new self($filename);
    }

    /**
     * Load the image
     */
    private function _load()
    {
        try {
            ini_set('memory_limit', '256M');
            $this->_image = \Intervention\Image\ImageManagerStatic::make($this->_path);
        } catch (\Exception $e) {
            throw new \Exception("Unable to load this image");
        }
    }

    /**
     * Save the image on the filesystem
     */
    private function _save()
    {
        $this->_image->save($this->_path);
    }

    /**
     * Get the path for the resized image
     *
     * @param Integer $width
     * @param Integer $height
     * @return String
     */
    private function _pathResized($width, $height)
    {
        $extension = pathinfo($this->_path, PATHINFO_EXTENSION);
        $basename = basename($this->_path, '.' . $extension);
        return self::_path(sprintf("%s-%dx%d.%s", $basename, $width, $height, $extension));
    }

    /**
     * Get the path for the cropped image
     *
     * @param Integer $x
     * @param Integer $y
     * @return String
     */
    private function _pathCropped($x, $y, $width, $height)
    {
        $extension = pathinfo($this->_path, PATHINFO_EXTENSION);
        $basename = basename($this->_path, '.' . $extension);
        return self::_path(sprintf("%s-%dx%d-%dx%d.%s", $basename, $width, $height, $x, $y, $extension));
    }

    /**
     * Get the complete path for a given filename
     *
     * @param String $filename
     * @return String
     */
    private static function _path($filename)
    {
        return __DIR__ . '/../tmp/' . $filename;
    }

    /**
     * Generate a filename
     *
     * @param String $path
     * @return String
     */
    private static function _filename($path)
    {
        $extension = pathinfo($path, PATHINFO_EXTENSION);
        return sprintf("%d.%s", time(), $extension);
    }

}
